validator/client/src/main/java/tech/pegasys/teku/validator/client/KeyManager.java

```diff
@@ -74,15 +83,65 @@ public KeyManager(final ValidatorLoader validatorLoader, final DataDirLayout dat
    */
-  public DeleteKeysResponse deleteValidators(final List<BLSPublicKey> validators) {
-    final List<DeleteKeyResult> deletionResults =
-        validators.stream()
-            .map(
-                key ->
-                    DeleteKeyResult.error(
-                        String.format("error: key %s not deleted", key.toAbbreviatedString())))
-            .collect(Collectors.toList());
-    return new DeleteKeysResponse(deletionResults, "");
+  public synchronized DeleteKeysResponse deleteValidators(final List<BLSPublicKey> validators) {
+    final List<DeleteKeyResult> deletionResults = new ArrayList<>();
+    final SlashingProtectionIncrementalExporter exporter =
+        new SlashingProtectionIncrementalExporter(
+            ValidatorClientService.getSlashingProtectionPath(dataDirLayout));
+    for (final BLSPublicKey publicKey : validators) {
+      Optional<Validator> maybeValidator =
+          validatorLoader.getOwnedValidators().getValidator(publicKey);
+
+      // read-only check in a non-destructive manner
+      if (maybeValidator.isPresent() && maybeValidator.get().isReadOnly()) {
+        deletionResults.add(DeleteKeyResult.error("Cannot remove read-only validator"));
+        continue;
+      }
+      // delete validator from owned validators list
+      maybeValidator = validatorLoader.getOwnedValidators().removeValidator(publicKey);
+      if (maybeValidator.isPresent()) {
+        deletionResults.add(deleteValidator(maybeValidator.get(), exporter));
+      } else {
+        deletionResults.add(attemptToGetSlashingDataForInactiveValidator(publicKey, exporter));
+      }
+    }
+    String exportedData;
+    try {
+      exportedData = exporter.finalise();
+    } catch (JsonProcessingException e) {
+      LOG.error("Failed to serialize slashing export data", e);
+      exportedData = EXPORT_FAILED;
+    }
+    return new DeleteKeysResponse(deletionResults, exportedData);
+  }
+
+  @VisibleForTesting
+  DeleteKeyResult attemptToGetSlashingDataForInactiveValidator(
+      final BLSPublicKey publicKey, final SlashingProtectionIncrementalExporter exporter) {
+    if (exporter.haveSlashingProtectionData(publicKey)) {
+      final Optional<String> error = exporter.addPublicKeyToExport(publicKey, LOG::debug);
+      return error.map(DeleteKeyResult::error).orElseGet(DeleteKeyResult::notActive);
+    } else {
+      return DeleteKeyResult.notFound();
+    }
+  }
+
+  @VisibleForTesting
+  DeleteKeyResult deleteValidator(
+      final Validator activeValidator, final SlashingProtectionIncrementalExporter exporter) {
+    final Signer signer = activeValidator.getSigner();
+    signer.delete();
+    LOG.info("Deleted validator: " + activeValidator.getPublicKey().toAbbreviatedString());
+    final DeleteKeyResult deleteKeyResult =
+        validatorLoader.deleteMutableValidator(activeValidator.getPublicKey());
+    if (deleteKeyResult.getStatus().equals(DeletionStatus.DELETED)) {
```

Comment:
`DeletionStatus` is an enum so can just use `==` here
```suggestion
    if (deleteKeyResult.getStatus() == DeletionStatus.DELETED) {
```